plugins/catalog-backend/src/ingestion/processors/ldap/vendors.ts

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2021 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { SearchEntry } from 'ldapjs';
+
+/**
+ * An LDAP Vendor handles unique nuances between different vendors.
+ */
+export type LdapVendor = {
+  /**
+   * The attribute name that holds the distinguished name (DN) for an entry.
+   */
+  dnAttributeName: string;
+  /**
+   * The attribute name that holds a universal unique identifier for an entry.
+   */
+  uuidAttributeName: string;
+  /**
+   * Decode ldap entry values for a given attribute name to their string representation.
+   *
+   * @param entry The ldap entry
+   * @param name The attribute to decode
+   */
+  decodeStringAttribute: (entry: SearchEntry, name: string) => string[];
+};
+
+export const DefaultLdapVendor: LdapVendor = {
+  dnAttributeName: 'entryDN',
+  uuidAttributeName: 'entryUUID',
+  decodeStringAttribute: (entry, name) => {
+    return decode(entry, name, value => {
+      return value.toString();
+    });
+  },
+};
+
+export const ActiveDirectoryVendor: LdapVendor = {
+  dnAttributeName: 'distinguishedName',
+  uuidAttributeName: 'objectGUID',
+  decodeStringAttribute: (entry, name) => {
+    const decoder = (value: string | Buffer) => {
+      if (name === ActiveDirectoryVendor.uuidAttributeName) {
+        return formatGUID(value);
+      }
+      return value.toString();
+    };
+    return decode(entry, name, decoder);
+  },
+};
+
+// Decode an attribute to a consumer
+function decode(
+  entry: SearchEntry,
+  attributeName: string,
+  decoder: (value: string | Buffer) => string,
+): string[] {
+  const values = entry.raw[attributeName];
+  if (Array.isArray(values)) {
+    return (values as Buffer[]).map(v => {
```

Comment:
this cast is a bit worrying and probably skippable - isn't the decoder supposed to handle both buffers and strings anyway?

Reply:
Yes, it an handle both Buffers and string so I have updated the PR to not cast the values